Feature/smd 16

hbgwhite opened the pull request:
Added Processors and Validators to check the entire XML document for violations of the specification located here: http://www.smarterapp.org/documents/AdministrationTestPackageFormat.pdf

hbgwhite committed:
Finished up all xml parsing/validation except admin segments and moved processors into folders that more closely match their xml hierarchy

hbgwhite committed:
- AdminSegmentProcessor
- ItemSelectionProcessor stub
- SegmentBlueprintProcessor stub
- SegmentFormProcessor stub
- SegmentPoolProcessor stub
- TestFormProcessor stub

hbgwhite committed:
- Updated AdminSegmentProcessor to include all sub-elements
- ItemSelectorProcessor
- SegmentBlueprintProcessor stub
- SegmentFormProcessor stub
- SegmentPoolProcessor stub
- TestFormProcessor stub
- ItemSelectionParameterProcessor
- ItemSelectionParameterPropertyProcessor
- ItemSelectorIdentifierProcessor

hbgwhite committed:
- RequiredBooleanValidator
- SegmentItemGroupIdentifierProcessor
- SegmentGroupItemProcessor
- SegmentBlueprintElementProcessor
- TestFormProcessor stub
- SegmentPoolProcessor
- SegmentFormProcessor
- SegmentBlueprintProcessor

Various updates to existing classes to include new processors in trees

hbgwhite committed:
- TestFormProcessor
- ItemGroupIdentifierProcessor
- TestFormIdentifierProcessor
- TestFormPartitionIdentifierProcessor
- TestFormPartitionProcessor
- TestFormPoolPropertyProcessor

hbgwhite committed:
- TestFormPartitionItemGroupProcessor
- GroupItemProcessor
- PassageRefProcessor

Full XML tree complete

hbgwhite committed:
Code Cleanup

Commit:
- Added TabulateSmarterTestScoringProject
- Applied formatting to Processors and Validators
- Moved redundant XPathNavigator property to base class in Processors

Commit:
Reverted .sln Visual Studio Version

Commit:
Example of new improved Validator format. Code location is in the new scoring package.

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/Validator.cs:
```diff
@@ -15,6 +15,11 @@ protected Validator(ErrorSeverity errorSeverity, object parameter = null)
 
+        public bool IsValid(object value, bool isRequired)
+        {
+            return IsValid(value) || !isRequired && value == null;
```
Can you add parenthesis around the logic so it's clear which are grouped together?  Just be explicit instead of having to remember the operator precedence

hbgwhite replied:
I'll drop the second conditional onto its own line. That's consistent with the formatting style i'm using elsewhere

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/ValidatorCollection.cs:
```diff
@@ -30,5 +30,10 @@ public bool IsValid(object value)
+
+        public bool IsValid(object value, bool isRequired)
+        {
+            return IsValid(value) || !isRequired && value == null;
```
Same here.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/AdminSegmentProcessor.cs:
```diff
@@ -0,0 +1,99 @@
+            return false;
+        }
+
+        //TODO: enum
```
Is this needed?  Seems left over

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/GroupItemProcessor.cs:
```diff
@@ -0,0 +1,182 @@
+        private static readonly XPathExpression sXp_ResponseRequired = XPathExpression.Compile("@responserequired");
+        internal static readonly XPathExpression sXp_IsActive = XPathExpression.Compile("@isactive");
+        internal static readonly XPathExpression sXp_IsFieldTest = XPathExpression.Compile("@isfieldtest");
+        private static readonly XPathExpression sXp_BlockId = XPathExpression.Compile("@blockid");
```
Is there a need for some to be internal and some private?

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/AdminSegmentProcessor.cs:
```diff
@@ -0,0 +1,99 @@
+                   && SegmentBlueprintProcessor.Process()
+                   && ItemSelectorProcessor.Process()
+                   && SegmentPoolProcessor.Process()
+                   && SegmentFormProcessor.Process();
```
If `IsValidSegmentId` (for example) returns false, then the Processors below and the other validation methods won't be called.  I'm guessing this will be a problem.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/GroupItemProcessor.cs:
```diff
@@ -0,0 +1,182 @@
+                   && IsValidBlockId();
+        }
+
+        internal bool IsValidItemId()
```
private would work for these methods right since they are only called by this class?

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/GroupItemProcessor.cs:
```diff
@@ -0,0 +1,182 @@
+            return false;
+        }
+
+        // Optional
```
If you are going to have this comment can you expand on it a bit so it's more clear.  I can see once looking at the code that the form position can be null and it is still considered valid, but was confused at first just reading the comment.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/ItemSelectionParameterProcessor.cs:
```diff
@@ -0,0 +1,52 @@
+
+        internal ItemSelectionParameterProcessor(XPathNavigator navigator) : base(navigator)
+        {
+            ItemSelectionParameterPropertyProcessors = new List<ItemSelectionParameterPropertyProcessor>();
```
You should be able to put this initialization as part of the auto-property now in C# 6.

```
private IList<ItemSelectionParameterPropertyProcessor> ItemSelectionParameterPropertyProcessors { get; } = new List<ItemSelectionParameterPropertyProcessor>();
```

hbgwhite replied:
That is nifty! Changing it.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/ItemSelectionParameterProcessor.cs:
```diff
@@ -0,0 +1,52 @@
+        public override bool Process()
+        {
+            return IsValidBpElementId()
+                   && ItemSelectionParameterPropertyProcessors.All(x => x.Process());
```
I think `.All()` will short circuit when it fails the first one so they might not all process.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/ItemSelectionParameterPropertyProcessor.cs:
```diff
@@ -0,0 +1,47 @@
+                new MaxLengthValidator(ErrorSeverity.Degraded, 100)
+            };
+            Value = Navigator.Eval(sXp_Value);
+            if (validators.IsValid(200))
```
This should be `validators.IsValid(Value)`, right?

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/ItemSelectionParameterPropertyProcessor.cs:
```diff
@@ -0,0 +1,47 @@
+                new MaxLengthValidator(ErrorSeverity.Degraded, 500)
+            };
+            Label = Navigator.Eval(sXp_Label);
+            if (validators.IsValid(200))
```
`validators.IsValid(Label)`

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/ItemSelectorProcessor.cs:
```diff
@@ -0,0 +1,57 @@
+            foreach (XPathNavigator selectionParameter in itemSelectionParameters)
+            {
+                ItemSelectionParameterProcessors.Add(new ItemSelectionParameterProcessor(selectionParameter));
+            }
```
Seems like line 20 - 25 could be put into a simple LINQ statement something like (not tested)

```
ItemSelectionParameterProcessors  = navigator
    .Select("itemselectionparameter")
    .Select(x => new ItemSelectionParameterProcessor(x))
    .ToList();
```

The double Select() feels weird, so maybe there is something a little cleaner, not sure.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdminSegment/SegmentBlueprintProcessor.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+            SegmentBlueprintElementProcessors = new List<SegmentBlueprintElementProcessor>();
+            var blueprintElements = navigator.Select("segmentbpelement");
+            foreach (XPathNavigator blueprintElement in blueprintElements)
```
similar LINQ as above would work

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/AdministrationProcessor.cs:
```diff
@@ -24,21 +27,30 @@ internal AdministrationProcessor(XPathNavigator navigator)
+            var testForms = navigator.Select("testform");
+            foreach (XPathNavigator testForm in testForms)
+            {
+                TestFormProcessors.Add(new TestFormProcessor(testForm));
```
same LINQ comment.  there are a lot of places like that it seems, so if that makes sense can you check all of them.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/ItemPool/TestItem/BPrefProcessor.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace TabulateSmarterTestAdminPackage.Processors.Specification.TestSpecification.Administration.ItemPool.TestItem
+{
+    internal class BPrefProcessor : Processor
```
BpRefProcessor

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/ItemPool/TestItem/ItemScoreParameterProcessor.cs:
```diff
@@ -0,0 +1,64 @@
+                   && IsValidValue();
+        }
+
+        // TODO: enum
```
What do you mean by enum?

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/ItemPool/TestItem/PassageRefProcessor.cs:
```diff
@@ -0,0 +1,38 @@
+                new MaxLengthValidator(ErrorSeverity.Benign, 100)
+            };
+            PassageRef = Navigator.Value;
+            if (validators.IsValid(PassageRef))
```
Should this be using the IsValid() method that has the 2nd parameter for isRequired?

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/PoolPropertyProcessor.cs:
```diff
@@ -1,27 +1,23 @@
-        private static readonly XPathExpression sXp_Label = XPathExpression.Compile("@label");
+        internal static readonly XPathExpression sXp_Property = XPathExpression.Compile("@property");
+        internal static readonly XPathExpression sXp_Value = XPathExpression.Compile("@value");
+        internal static readonly XPathExpression sXp_Label = XPathExpression.Compile("@label");
```
why internal?

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/TestBlueprint/BlueprintElementProcessor.cs:
```diff
@@ -76,13 +75,14 @@ internal bool IsValidParentId()
             };
-            ParentId = _navigator.Eval(sXp_ParentId);
+            ParentId = Navigator.Eval(sXp_ParentId);
             if (ParentId == null || validators.IsValid(ParentId))
```
the new IsValid() with 2 parameters should be used here

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/Administration/TestBlueprint/BlueprintElementProcessor.cs:
```diff
@@ -186,13 +191,14 @@ internal bool IsMinFtItemsLessThanMaxFtItems()
-            if ((string.IsNullOrEmpty(MinFtItems)
-                 && string.IsNullOrEmpty(MaxFtItems))
+            if (string.IsNullOrEmpty(MinFtItems)
+                && string.IsNullOrEmpty(MaxFtItems)
```
Can you put the parenthesis back in place for a more obvious grouping

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/PropertyProcessor.cs:
```diff
@@ -1,26 +1,22 @@
-        private string Value { get; set; }
-        private string Label { get; set; }
+        internal string Value { get; set; }
+        internal string Label { get; set; }
```
is internal needed?

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Specification/TestSpecification/PropertyProcessor.cs:
```diff
@@ -68,12 +66,13 @@ internal bool IsValidLabel()
             };
-            Label = _navigator.Eval(sXp_Label);
+            Label = Navigator.Eval(sXp_Label);
             if (validators.IsValid(200))
```
`validators.IsValid(Label)`

Commit:
- Updated .NET version for all projects
- Moved and renamed the ReportingUtility
- Moved PropertyProcessor to common
- Updated processor superclass

Commit:
- Added ConvenienceValidators
- Refactored existing classes to adapt to new pattern

Commit:
Error validation for TestSpecificationProcessor "publisher" field

Commit:
Refactor except AdminSegments and TestForms

Commit:
Refactored AdminPackageTabulator to use new validation format

Commit:
Applied rename to class per JT's pull request comments
